tools/check_links.py

```diff
@@ -31,37 +32,57 @@
 LINK = collections.namedtuple('LINK', 'dest valid')
 
 
-def check_docs(dir_name):
-  'Traverse dir_name and check links in Markdown files.'
+def check_link(link, readme_path, external):
+  'Checks if a link element has a valid destination.'
+  link_valid = None
+  url = urllib.parse.urlparse(link.dest)
+  # If the link is public, say the link is anyway valid
+  # if --external is not set; check the link otherwise
+  if url.scheme:
+    link_valid = True
+    if external:
+      try:
+        response = requests.get(link.dest)
+        link_valid = response.ok
+      except requests.exceptions.RequestException:
+        link_valid = False
+  # The link is private
+  else:
+    link_valid = (readme_path.parent / url.path).exists()
+  return LINK(link.dest, link_valid)
+
+
+def check_docs(dir_name, external=False):
+  'Traverses dir_name and checks for all Markdown files.'
   dir_path = BASEDIR / dir_name
+  parser = marko.parser.Parser()
   for readme_path in sorted(dir_path.glob('**/*.md')):
     if '.terraform' in str(readme_path) or '.pytest' in str(readme_path):
       continue
+
+    root = parser.parse(readme_path.read_text())
+    elements = collections.deque([root])
```

Comment:
Let me restate: iterating beats creating lists or list-like. Can you please iterate over children instead of assembling a list?